### ISSUE TITLE
Customer function failing with socket exceptions and "unable to retrieve master key"

### ISSUE THREAD

**Issue by brettsam** [Azure/azure-webjobs-sdk-script#1621]
Using this to track customer issue: https://twitter.com/mpavlak12/status/877672738676244480

**Comment by brettsam** [comment]
It looks like the app is hitting the 300-connection limit and failing to make any new connections. This is causing the host to fail at startup, which would explain a bunch of issues. It'd help to know what the functions are doing and (optimally) see the code where calls to external services are being made.

**Comment by spelltwister** [comment]
Here is the last thing I tried to run when it failed.

    public class JsonBlobber
    {
        protected CloudStorageAccount StorageAccount { get; }
        protected Lazy<CloudBlobClient> BlobClient { get; }
        protected Lazy<CloudBlobContainer> BlobContainer { get; }
        public JsonBlobber(string connectionString, string containerName) : this(CloudStorageAccount.Parse(connectionString), containerName) { }
        public JsonBlobber(CloudStorageAccount storageAccount, string containerName)
        {
            this.StorageAccount = storageAccount;
            this.BlobClient = new Lazy<CloudBlobClient>(() => this.StorageAccount.CreateCloudBlobClient());
            this.BlobContainer = new Lazy<CloudBlobContainer>(() =>
            {
                var container = this.BlobClient.Value.GetContainerReference(containerName);
                container.CreateIfNotExists(BlobContainerPublicAccessType.Off);
                return container;
            });
        }

        public Task SaveResultsAsync(object data, string dataId)
        {
            var jsonData = JsonConvert.SerializeObject(data);
            var blob = this.BlobContainer.Value.GetBlockBlobReference($"{dataId}");
            return blob.UploadTextAsync(jsonData);
        }

        public IEnumerable<Task<T>> GetResultsPrefixedWhereAsync<T>(string prefix, Func<CloudBlockBlob, bool> blobFilter)
        {
            return this.BlobContainer
                .Value
                .ListBlobs(prefix, true, BlobListingDetails.Metadata)
                .OfType<CloudBlockBlob>()
                .Where(blobFilter)
                .Select(async x => JsonConvert.DeserializeObject<T>(await x.DownloadTextAsync()));
        }

        public IEnumerable<Task<T>> GetResultsPrefixedAsync<T>(string prefix)
        {
            var blobs = this.BlobContainer
                .Value
                .ListBlobs(prefix, true);

            return blobs.OfType<CloudBlockBlob>().Select(async x => JsonConvert.DeserializeObject<T>(await x.DownloadTextAsync()));
        }
    } 


    public static class Aggregator
    {
        private static readonly Lazy<JsonBlobber> JsonExploreBlobberLazy = new Lazy<JsonBlobber>(() => new JsonBlobber(ConfigurationManager.ConnectionStrings["storage"].ConnectionString, "containerName"));
        private static JsonBlobber Blobber => JsonExploreBlobberLazy.Value;

        [NoAutomaticTrigger]
        [FunctionName("RebuildAggregations")]
        public static async Task Rebuild(string unused, TraceWriter log)
        {
            var results = await Task.WhenAll(Blobber.GetResultsPrefixedAsync<DataObject>("imports")).ConfigureAwait(false);
            var aggregations = CreateAggregationsByFriendlyId(results);
            await Task.WhenAll(aggregations.Select(x => Blobber.SaveResultsAsync(x, $"agg/{x.FriendlyId}"))).ConfigureAwait(false);
        }

        private static TransformedObject[] CreateAggregationsByFriendlyId(DataObject[] datas)
        {
            return datas.GroupBy(x => x.Name.UrlFriendlyString(), StringComparer.OrdinalIgnoreCase)
                        .Select(x =>
                            new TransformedObject()
				{
					// propery mapping and aggregation, just more linq to objects, no external calls
				};
                        )
                        .ToArray();
        }
    }

**Comment by spelltwister** [comment]
Prior to that, I was running this function to transform and localize assets.  Most data objects have no images.  That ones that do typically have 1.  More than 2 is extremely rare.  This function should have run about 5,000 times in parallel.

        public class ImportBlobberBase
        {
            protected JsonBlobber DataBlobber { get; }
            protected ImageBlobber ImageBlobber { get; }
            protected HttpClient Client { get; }

            public ImportBlobberBase(string connectionString, string containerName, HttpClient client)
                : this(CloudStorageAccount.Parse(connectionString), containerName, client)
            {
            }

            public ImportBlobberBase(CloudStorageAccount storageAccount, string containerName, HttpClient client)
            {
                this.DataBlobber = new JsonBlobber(storageAccount, containerName);
                this.ImageBlobber = new ImageBlobber(storageAccount, containerName);
                this.Client = client;
            }

            protected async Task LocalizePhotos(DataObject data, string imageNamePrefix)
            {
                data.Images = (await Task.WhenAll(data.Images.Select(async x =>
                {
                    byte[] image;
                    try
                    {
                        // a bunch of the images return 403 response, just ignore
                        image = await DownloadImageAsync(x.PhotoUrl).ConfigureAwait(false);
                    }
                    // TODO: 403
                    catch
                    {
                        return null;
                    }
                    var uri = await this.ImageBlobber.SaveImageAsync(image, $"{imageNamePrefix}/{x.Id}").ConfigureAwait(false);
                    x.PhotoUrl = uri.AbsoluteUri;
                    return x;
                })).ConfigureAwait(false))
                .Where(x => x != null)
                .ToArray();
            }

            public async Task<byte[]> DownloadImageAsync(string imageUrl)
            {
                var imgResponse = await this.Client.GetAsync(imageUrl).ConfigureAwait(false);
                imgResponse.EnsureSuccessStatusCode();
                var imgBytes = await imgResponse.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
                return imgBytes;
            }
        }
      
      public class SpecializedImportBlobber : ImportBlobberBase
        {
            public SpecializedImportBlobber(string connectionString, string containerName, HttpClient client)
                : base(connectionString, containerName, client)
            {
            }

            public async Task<TransformedObject> Save(DataObject data)
            {
                TransformedObject transformedObject = SpecializedConverter.Convert(data);

                if (transformedObject == null)
                {
                    return null;
                }

                string namePrefix = $"imports/{data.id}";
                await LocalizePhotos(data, namePrefix).ConfigureAwait(false);

                await this.DataBlobber.SaveResultsAsync(transformedObject, $"{namePrefix}/convertedData").ConfigureAwait(false);

                return transformedObject;
            }
        }
    }
    
    public static class Transform
    {
        private static readonly Lazy<JsonBlobber> BlobberLazy = new Lazy<JsonBlobber>(() => new JsonBlobber(ConfigurationManager.ConnectionStrings["storage"].ConnectionString, "containerName"));
        private static JsonBlobber Blobber => BlobberLazy.Value;

        private static readonly Lazy<SpecializedImportBlobber> ImportBlobberLazy = new Lazy<SpecializedImportBlobber>(() => new SpecializedImportBlobber(ConfigurationManager.ConnectionStrings["storage"].ConnectionString, "containerName", new HttpClient()));
        private static SpecializedImportBlobber ImportBlobber => ImportBlobberLazy.Value;

        [FunctionName("Importer")]
        public static async Task Run([BlobTrigger("data/{id}", Connection = "storage")]string myBlob, string id, TraceWriter log)
        {
            var data = JsonConvert.DeserializeObject<DataObject>(myBlob);
            var transformedObject = await ImportBlobber.Save(data).ConfigureAwait(false);
            await Blobber.SaveResultsAsync(transformedObject, $"imports/{transformedObject.Name.UrlFriendlyString()}${transformedObject.SourceId}").ConfigureAwait(false);
        }
    }

**Comment by spelltwister** [comment]
Prior to that was another function that downloads data from a web api same as the above one and saves into storage.  There would be 5,000 parallel executions of that as well.  However, once they complete, doesn't any resource used get freed up or how is something sticking around.  anything i can do to clean up the sockets?

**Comment by spelltwister** [comment]
I created a new service plan and deployed the functions again.  I was able to manually trigger the first bit of code that does not read at all external, only reads/write blobs, but it failed with the socket exceptions still when reading the list of blobs and downloading the contents.

**Comment by brettsam** [comment]
The issue is with how you're firing of so many Tasks to download blobs at once -- if there's 5,000 each of those will spawn up a new connection to the blob service. The consumption plan runs within a sandbox (limits here: https://github.com/projectkudu/kudu/wiki/Azure-Web-App-sandbox#per-sandbox-per-appper-site-numerical-limits) so you'll hit that 300-connection limit pretty fast like this. 

Some options:
1. Slow down the requests by waiting for each one to complete before continuing. That will obviously take longer, and may bump up against the 10-minute limit that we have for Consumption functions, but you'd limit your connections. 
2. Similar to above -- but use the [DataFlow library](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/dataflow-task-parallel-library) to batch up your tasks into bunches of 20 or 50 or 100 or so -- that'd be faster than doing each one at a time and should keep you below the connection limit.
3. Depending on how you want to trigger these (is it manually)? You could have something like a BlobTrigger  function whose job it is to take one blob at a time and append the relevant data to a larger AppendBlob (or put data in Table storage or DocDb, etc -- basically copy it elsewhere for faster retrieval). Then you have an Aggregate function whose job it is to operate on that single, aggregated file (or table, etc). Those copying functions would each run quickly (and allow for better scale-out), and downloading the single batch of data to aggregate would be faster.

I know the connection limit can be a rough thing to bump into. This is an interesting scenario that I think we'd like to make easier in the future -- feedback on how you end up accomplishing it would be great. Thanks for reaching out to us.

**Comment by spelltwister** [comment]
There are 5000 blobs, but each function handles one of them.  In the first case, yes, I download all the blobs.  I will switch the code to instead pump the names into a queue so that the other functions can each handle one.  However, why do the connections affect the other functions and why don't the connections close after the function fails?  Seems like once the error happens, I'm dead in the water forever.

I will update the code as indicated to use a queue and see if that helps things.

**Comment by brettsam** [comment]
All functions in a function app run within a single host, and the connections are watched based on the host. When you scale out, the whole host is scaled.

I believe that connections are kept open for ~2 minutes after they're created in another request is made. If you run one-after-the-other, they should all use the same connection, but when you run all at once, new ones are created because the others are in use. My guess is that you're running into this delay before connections are closed. I think if you restarted your site via the portal, the connections would reset and you'd be able to try it again (although with 5000 connections you'd hit the limit again).

**Comment by spelltwister** [comment]
![image](https://user-images.githubusercontent.com/4492581/27501856-8a5c2d64-585f-11e7-89e3-d2e873e71296.png)

every function app I create has all management options disabled as seen above.  I cannot stop, delete, swap, restart, ...

**Comment by brettsam** [comment]
You may be hitting this: https://github.com/Azure/azure-functions-ux/issues/1469#issuecomment-309130468

**Comment by spelltwister** [comment]
Following the linked steps enabled the disabled options.

I updated the function to pump messages into a queue using the ICollector<string> parameter instead of doing all the work and it's working nicely.

Thanks for the help!

**Comment by spelltwister** [comment]
While the buttons are enabled, it will not actually delete the app.  I tried several times and it says it is working on it, but never completes.

**Comment by spelltwister** [comment]
And I ran into the problem again.  I do not know how I'm creating too many connections as everything I do is batched now.  I even reduced the batch size to 2 items and it couldn't do it.  To get around that, I associated the function with an app service plan running a website.  I would like to eventually figure that out, but at least I have the option for now.

**Comment by brettsam** [comment]
Would you mind sharing your latest code?

**Comment by spelltwister** [comment]
I will try to trim down the code to highlight what the problem is.  The code running into the problem has a ton of steps, but only 2 which use async calls to external services.  I put both blocks of code into 

    foreach(var batch in tasks.Chunk(2)) { await Task.WhenAll(batch).ConfigureAwait(false) }

 and it still couldn't handle it.  I'm not sure how it counts open connections when scaled out.  Each function operates on a single blob and should not make more than 4 or so calls at a time using the chunking. :(

**Comment by spelltwister** [comment]
I am still not able to delete the function.  it just sits forever trying to delete.  Been few days now with status "Running" in my portal alerts.

**Comment by brettsam** [comment]
I'm going to create a new issue for the delete problem since it sounds unrelated: https://github.com/Azure/Azure-Functions/issues/371

In that issue, could you share your function app name -- either directly or [indirectly](https://github.com/projectkudu/kudu/wiki/Reporting-your-site-name-without-posting-it-publicly)? We can look at our logs to see what may be the problem.
